### ISSUE TITLE
Two intercooler attributes are interfering with each other

### ISSUE THREAD

**Issue by qbpd** [LeadDyno/intercooler-js#223]
I'm working on a shopping cart-like feature where there is a grid of pictures, with each picture having a little shopping cart icon on top.

This grid is using intercooler infinite scroll, which is a combination of append-from/scrolled-into-view that was one of the provided examples.

```
<div class="itemContainer">
    <div class="itemImage">...</div>
    <div class="itemImage">...</div>
    (small handful of these images are loaded)
    <div class="itemImage">...</div>
    <div ic-append-from= (my get path, using pagination)
         ic-trigger-on="scrolled-into-view"
         ic-trigger-delay="150ms"
         ic-target="#targetDiv"
         ic-indicator="#spinner">
    </div>
</div>
```

As the user scrolls towards the bottom, a div like above gets added to #targetDiv so it will eventually be organized like:

```
<div id="#targetDiv">
    <div class="itemContainer">....</div>
    <div class="itemContainer">....</div>
    <div class="itemContainer">....</div>
    ...(as user scrolls, a single itemContainer gets added)
</div>
```

When you click on the shopping cart icon, ic-put-to gets called, and the server returns a different shopping-cart-icon to replace the last one

```
<div class="itemImage">
    ...
    <div class="cartdiv" id=(item_id)
        <a ic-put-to= (PUT url)
           ic-replace-target=(#item_id from above)>
            <img class=("is_incart"/"is_not_incart")
                  src=(image url) />
        </a>
    </div>
</div>
```

On first glance, everything appeared to work fine, but I quickly noticed there's a show-stopping bug.

When I click the cart icon, on top of the expected PUT, there is one GET call made for each "ic-append-from" div in the DOM. So if I scroll for a while, and click on a cart icon, that would result in 1 PUT and many GET requests being made at once. These GET requests are the same kind as the paginated requests that are made when you scroll-into-view.

So in other words, if I have 3 itemContainers in the DOM, click the cart icon, the PUT request leads to the replacing of my cart image, and the 3 GET requests lead to me having 6 itemContainers (the second 3 being a duplicate of the first 3). The next click would lead to 1 PUT and 12 GET requests

The infinite scrolling and cart-icon parts appear to work perfectly fine when by themselves -- It's when they are paired together in my scenario above that I am noticing issues. When I remove the div containing "ic-append-from", this issue no longer exists (although then I do not have infinite scrolling).

I have spent a couple of hours last night trying to debug the javascript and have jotted down some observations that might help, but to be honest, dealing with JS is something I'm really weak at and I'd like for someone to take a look at this with a fresh set of eyes first or at least help point me in the right direction.

There are no other javascript files/JS libraries involved. I am using Elixir/Phoenix templates to go along with intercooler 1.1.2

This has been a wonderful tool for me so far. Thank you for listening to my problem.

**Comment by carsongross** [comment]
I will try to reproduce this today

**Comment by carsongross** [comment]
OK, I tried to reproduce the issue and was not able to locally.  I think the problem is that somehow the intercooler attributes on the row are interfering with the inner links.

Can you get into a situation where you are seeing the multiple GETs being issued and send me the source of the document at that point?  (run document.documentElement.innerHTML from the console)

As a short term work around, while we get to the bottom of the issue, you might try putting the element that has 'scrolled-into-view' on it so it is a sibling, rather than a parent element of the elements that are being clicked.

Finally, can you tell me what browser you are using?

**Comment by qbpd** [comment]
Thank you for looking into this! I will send you the document source and then attempt that work-around when I get back home. This issue was experienced with the latest Chrome and Safari on OSX (Sierrra).

**Comment by carsongross** [comment]
OK, I was on OSX/Chrome, so if it is a bug it isn' the obvious one.

**Comment by qbpd** [comment]
I've attached the source of the document after the multiple GETs were issued.

Here were the network calls made to get to this point:

http://localhost:4001/users/test/items?page=2&ic-request=true&&ic-id=7&ic-target-id=targetDiv&ic-current-url=%2Fsearch&_method=GET

(after scrolling to the bottom of the page, before the PUT request)

http://localhost:4001/users/test/items/789?incart=false&_csrf_token=OhU7ERcFPycRNyoMRC4iChQFLQc7EAAAwQXaZPrpgnre%2FWkDE1ealw%3D%3D
http://localhost:4001/users/test/items?page=2&ic-request=true&&ic-id=7&ic-target-id=targetDiv&ic-current-url=%2Fsearch&_method=GET
http://localhost:4001/users/test/items?page=3&ic-request=true&&ic-id=7&ic-target-id=targetDiv&ic-current-url=%2Fsearch&_method=GET

[ic-interference.txt](https://github.com/LeadDyno/intercooler-js/files/1400791/ic-interference.txt)

**Comment by qbpd** [comment]
When I click on a cart icon again, the 4 requests were:
?page=2, ?page=3, ?page=3, and ?page=4.

And when clicking again, the 8 requests were: 2,3,3,4,3,5,4,4

Also, I tried playing around with the DOM to have the "div ic-append-from" element siblings with "cartdiv" and also siblings with "a ic-put-to". Neither serve as a work-around for me. Also, while playing with the DOM, I confirmed that removing these "ic-append-from" elements from the DOM has a direct impact on the number of GET requests

**Comment by carsongross** [comment]
HA.  I know what it is!

It is intercooler dependencies:

http://intercoolerjs.org/docs.html#dependencies

The problem is you are using a PUT, which is a mutation operation in the REST world, to a path (e.g. /users/test/items/234) that the paging path (/users/test/items) depends on, so intercooler automatically refreshes the parent path.

:)

I can think of two solutions here: 

# You can use a GET rather than a PUT for your image loading.  I think this is the the "right thing".  If you are using the HTTP verb to determine if it is an intercooler request, you can use the intercooler headers instead.
# You could add ic-deps="ignore" to the paging element.  It's a bit of a hack, but it would be the smallest change that would fix the issue.

**Comment by qbpd** [comment]
Thank you so much!!!

Oddly enough, I did skim the dependencies section before starting a ticket. I just...didn't understand it/didn't think it applied to my situation.

I went with your second solution because I was so eager to try it out (and I am still very excited), but also because I am modifying information (incart=true vs incart=false).

Thank you again!!! Have a great weekend :)
